botocore/endpoints_v2.py

```diff
@@ -0,0 +1,1123 @@
+        braces, and return the nested data from the rule's scoped variables. If the
+        string contains double curly braces such as `{{foo}}` they are treated as
+        the string literal `{foo}` and are not formatted.
+
+        :type value: str
+        :param value: A template string such as "https://{Bucket}.ec2.{url#authority}".
+
+        :type scope_vars: dict
+        :param scope_vars: A dictionary combining input parameters and variables that
+        have been scoped throughout the rule's evaluation.
+
+        :returns: A fully resolved string such as "https://us-west-2.ec2.amazonaws.com".
+        """
+        string_formatter = Formatter()
+        result = ""
+        for literal, reference, _, _ in string_formatter.parse(value):
+            if reference is not None:
+                template_value = scope_vars
+                template_params = reference.split("#")
+                for param in template_params:
+                    template_value = template_value[param]
+                result += f"{literal}{template_value}"
+            else:
+                result += literal
+        return result
+
+    def convert_func_name(self, value):
+        """Given a function signature name found in the `fn` key of a function
+        object, convert it to snake case with a leading underscore. Additionally
+        replace a period `.` with an underscore `_`. For example, `aws.partition`
+        becomes `_aws_partition` and `booleanEquals` becomes `_boolean_equals`.
+
+        :type value: str
+        :param value: A string value representing the function name.
+
+        :returns: The actual name of a function defined in this module.
+        """
+        change_case = f"_{xform_name(value)}"
+        return change_case.replace(".", "_")
+
+    def call_function(self, func_signature, scope_vars):
+        """Provided a function object, iterate through its arguments with
+        `resolve_value` and recurse if any of them are a nested function
+        object. If an arg is a reference object, dereference it by looking
+        for it in the rule's scoped variables. Finally, retrieve the function
+        and call it with the resolved arguments. If an `assign` parameter is
+        present, add the returned value to the rule's scoped variables for
+        the next condition/subordinate rule.
+
+        :type func_signature: dict
+        :param func_signature: A function object such as
+        {
+            "fn": "not",
+            "argv": [
+                {
+                    "fn": "isSet",
+                    "argv": [
+                        {
+                            "fn": "parseURL",
+                            "argv": [
+                                {
+                                    "ref": "Endpoint"
+                                }
+                            ]
+                        }
+                    ]
+                }
+            ]
+        }
```

Comment:
Rather than having these inline examples, let's add a module comment with a pointer to the vanilla tests directory. That will point any future developers where to look, and ensure we're always look at the most up-to-date examples.

Reply:
I wrote a comment that points both to the actual rules and test rules